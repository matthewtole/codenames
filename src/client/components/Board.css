.BoardWrapper {
  flex: 1;
}

.Board {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  padding: 1px;
}

.BoardRow {
  display: flex;
  flex-direction: row;
  height: 20%;
}
