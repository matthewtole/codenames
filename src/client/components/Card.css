.Card {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  width: 20%;
  background-color: #eee;
  font-size: 1.75rem;
  margin: 1px;
  user-select: none;
}

.Card--revealed {
  text-decoration: line-through;
  opacity: 0.2;
}

.Card--controller[data-role="blue"] {
  background-color: #bbdefb;
}

.Card--controller[data-role="red"] {
  background-color: #ef9a9a;
}

.Card--controller[data-role="bystander"] {
  background-color: #ff9100;
}

.Card--controller[data-role="assassin"] {
  background-color: #999;
}

.Card[data-role="blue"].Card--revealed {
  background-color: #2196f3;
  color: #bbdefb;
}

.Card[data-role="red"].Card--revealed {
  background-color: #d32f2f;
  color: #ef9a9a;
}

.Card[data-role="bystander"].Card--revealed {
  background-color: #ff9100;
}

.Card[data-role="assassin"].Card--revealed {
  background-color: #222;
}

.Card--highlighted {
  border: 5px solid #000;
}

@media only screen and (max-width : 1024px) {
    .Card {
        font-size: 1.2rem;
    }
}
