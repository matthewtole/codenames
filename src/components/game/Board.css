.BoardWrapper {
  flex: 1 1;
}

.Board {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  padding: 5px;
}

.BoardRow {
  display: flex;
  flex-direction: row;
  height: 20%;
}

@media only screen and (max-width: 1024px) {
  .Board {
    padding: 3px;
  }
}
