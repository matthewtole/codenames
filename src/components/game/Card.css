.Card {
  width: 20%;
  font-size: 1.75rem;
  /* margin: 1px; */
  user-select: none;
  position: relative;
}

.Card > div {
  background-color: #eee;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  top: 0;
  border-radius: 4px;
  transition: margin 0.1s ease-out;
  margin: 0;
  border: 2px solid #fff;
}

.Card--revealed {
  text-decoration: line-through;
  opacity: 0.2;
}

.Card--controller[data-role='BLUE_SPY'] > div {
  background-color: #bbdefb;
}

.Card--controller[data-role='RED_SPY'] > div {
  background-color: #ef9a9a;
}

.Card--controller[data-role='BYSTANDER'] > div {
  background-color: #ff9100;
}

.Card--controller[data-role='ASSASSIN'] > div {
  background-color: #999;
}

.Card[data-role='BLUE_SPY'].Card--revealed > div {
  background-color: #2196f3;
  color: #bbdefb;
}

.Card[data-role='RED_SPY'].Card--revealed > div {
  background-color: #d32f2f;
  color: #ef9a9a;
}

.Card[data-role='BYSTANDER'].Card--revealed > div {
  background-color: #ff9100;
}

.Card[data-role='ASSASSIN'].Card--revealed > div {
  background-color: #222;
}

.Card--highlighted > div {
  /* border: 5px solid #000; */
  /* box-shadow: 0 0 20px rgba(0, 0, 0, 0.5); */
  box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
  margin: -7px;
  z-index: 200;
}

@media only screen and (max-width: 1024px) {
  .Card {
    font-size: 1.2rem;
  }
}
