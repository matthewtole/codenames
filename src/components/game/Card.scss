@import '../../sass/_variables';

.Card {
  width: 20%;
  font-size: 1.75rem;
  user-select: none;
  position: relative;

  & > div {
    background-color: #eee;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    position: absolute;
    left: 0;
    right: 0;
    bottom: 0;
    top: 0;
    border-radius: 4px;
    margin: 5px;
    border: 2px solid #ddd;

    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
    transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);
  }

  &--revealed > div {
    border: 0;
  }

  &--highlighted > div {
    box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
    // margin: -7px;
    z-index: 10;
  }

  &--revealed {
    text-decoration: line-through;
    opacity: 0.2;
  }

  &[data-role='BLUE_SPY'] {
    &.Card--controller,
    &.Card--game-over {
      > div {
        background-color: $blue-spy--light;
        border-color: $blue-spy--dark;
      }

      &.Card--controller.Card--highlighted > div {
        background-color: $blue-spy--dark;
        color: #fff;
      }
    }

    &.Card--revealed {
      > div {
        background-color: $blue-spy--dark;
        color: $blue-spy--light;
      }
    }
  }

  &[data-role='RED_SPY'] {
    &.Card--controller,
    &.Card--game-over {
      > div {
        background-color: $red-spy--light;
        border-color: $red-spy--dark;
      }
    }

    &.Card--controller.Card--highlighted > div {
      background-color: $red-spy--dark;
      color: #fff;
    }

    &.Card--revealed {
      > div {
        background-color: $red-spy--dark;
        color: $red-spy--light;
      }
    }
  }

  &[data-role='BYSTANDER'] {
    &.Card--controller,
    &.Card--game-over {
      > div {
        background-color: $bystander--light;
        border-color: $bystander--dark;
      }
    }

    &.Card--controller.Card--highlighted > div {
      background-color: $bystander--dark;
      color: #fff;
    }

    &.Card--revealed {
      > div {
        background-color: $bystander--dark;
      }
    }
  }

  &[data-role='ASSASSIN'] {
    &.Card--controller,
    &.Card--game-over {
      > div {
        background-color: $assassin--light;
        border-color: $assassin--dark;
      }
    }

    &.Card--controller.Card--highlighted > div {
      background-color: $assassin--dark;
      color: #fff;
    }

    &.Card--revealed {
      > div {
        background-color: $assassin--dark;
      }
    }
  }
}

@media only screen and (max-width: 1024px) {
  .Card {
    font-size: 1.2rem;

    & > div {
      margin: 3px;
    }
  }
}
