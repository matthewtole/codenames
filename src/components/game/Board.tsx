import * as React from 'react';
import './Board.css';
import { Card as CardComponent } from './Card';
import {
  Card,
  Coordinate,
  Role,
  CoordinateValue,
} from '../../redux/game/types';

export enum BoardMode {
  VIEWER = 'VIEWER',
  CONTROLLER = 'CONTROLLER',
}

export interface BoardProps {
  width: number;
  height: number;
  cards: Card[];
  revealedCards: number[];
  mode: BoardMode;
  highlighted?: Coordinate;
  highlightedRow?: number;
  isGameOver: boolean;
  onRevealCard: (card: Coordinate) => void;
  onHighlightCard: (card?: Coordinate) => void;
}

export class Board extends React.PureComponent<BoardProps, {}> {
  constructor(props: BoardProps) {
    super(props);
    this.state = {};
  }

  render() {
    return (
      <div className="BoardWrapper" onClick={() => this.highlightCard()}>
        <div className="Board">
          {this.renderRowHighlight()}
          {Array(this.props.height)
            .fill(0)
            .map((_: number, row: CoordinateValue) => this.renderRow(row))}
        </div>
      </div>
    );
  }

  private renderRowHighlight() {
    if (this.props.highlightedRow === undefined) {
      return null;
    }
    const row = this.props.highlightedRow;
    return (
      <div className="Board__row-highlight-wrapper">
        <div
          className="Board__row-highlight"
          style={{ top: 0, height: `${row * 20}%` }}
        />
        <div
          className="Board__row-highlight"
          style={{
            top: `${(row + 1) * 20}%`,
            height: `${(5 - row + 1) * 20}%`,
          }}
        />
      </div>
    );
  }

  private renderRow = (row: CoordinateValue) => {
    return (
      <div className="BoardRow" key={`row_${row}`}>
        {Array(this.props.width)
          .fill(0)
          .map((__: number, col: CoordinateValue) => this.renderCard(row, col))}
      </div>
    );
  }

  private renderCard = (row: CoordinateValue, col: CoordinateValue) => {
    const index = this.coordinateToIndex({ row, col });
    return (
      <CardComponent
        key={`card_${row}_${col}`}
        word={this.getWord(index)}
        role={this.getRole(index)}
        revealed={this.isRevealed(index)}
        highlighted={this.isHighlighted({ row, col })}
        onHighlight={() => {
          this.highlightCard({ row, col });
        }}
        onReveal={() => {
          this.revealCard({ row, col });
        }}
        boardMode={this.props.mode}
        isGameOver={this.props.isGameOver}
      />
    );
  }

  private getWord = (index: number): string => {
    return this.props.cards[index].word;
  }

  private getRole = (index: number): Role => {
    return this.props.cards[index].role;
  }

  private isRevealed = (index: number): boolean => {
    return this.props.revealedCards.indexOf(index) >= 0;
  }

  private isHighlighted = (card: Coordinate): boolean => {
    if (this.props.highlighted === undefined) {
      return false;
    }
    return (
      this.props.highlighted.row === card.row &&
      this.props.highlighted.col === card.col
    );
  }

  private highlightCard = (card?: Coordinate) => {
    this.props.onHighlightCard(card);
  }

  private revealCard = (card: Coordinate) => {
    this.props.onRevealCard(card);
  }

  private coordinateToIndex(coordinate: Coordinate): number {
    return coordinate.row * this.props.width + coordinate.col;
  }
}
