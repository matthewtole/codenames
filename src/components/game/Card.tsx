import * as React from 'react';
import './Card.css';
import { Role } from '../../lib/types';
import * as cx from 'classnames';
import { BoardMode } from './Board';

export interface CardProps {
  word: string;
  role: Role;
  revealed: boolean;
  highlighted: boolean;
  boardMode: BoardMode;
  onReveal: () => void;
  onHighlight: () => void;
}

export class Card extends React.PureComponent<CardProps, object> {
  render() {
    return (
      <div
        className={cx({
          Card: true,
          'Card--highlighted': this.props.highlighted,
          'Card--revealed': this.props.revealed,
          [`Card--${this.props.boardMode}`]: true,
        })}
        data-role={this.props.role}
        onClick={this.handleClick}
      >
        <div>{this.props.word}</div>
      </div>
    );
  }

  handleClick = (e: React.MouseEvent<HTMLDivElement>) => {
    e.stopPropagation();
    if (this.props.revealed) {
      return;
    }
    if (this.props.highlighted) {
      this.props.onReveal();
    } else {
      this.props.onHighlight();
    }
  }
}
