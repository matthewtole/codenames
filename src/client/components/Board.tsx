import * as React from 'react';
import './Board.css';
import { Card } from './Card';
import { CardData, Coordinate, Role } from '../../shared/game';

export enum BoardMode {
  Viewer = 'viewer',
  Controller = 'controller',
}

export interface BoardProps {
  width: number;
  height: number;
  cards: CardData[];
  mode: BoardMode;
  highlighted?: Coordinate;
  onRevealCard: (card: Coordinate) => void;
  onHighlightCard: (card?: Coordinate) => void;
}

export interface BoardState {
  highlighted?: Coordinate;
}

export class Board extends React.Component<BoardProps, BoardState> {
  constructor(props: BoardProps) {
    super(props);
    this.state = {};
  }

  render() {
    const rows = [];
    for (let r = 0; r < this.props.height; r += 1) {
      const row = [];
      for (let c = 0; c < this.props.width; c += 1) {
        const card = { row: r, col: c };
        row.push((
          <Card
            key={`card_${r}_${c}`}
            word={this.getWord(card)}
            role={this.getRole(card)}
            revealed={this.isRevealed(card)}
            highlighted={this.isHighlighted(card)}
            onHighlight={() => { this.highlightCard(card); }}
            onReveal={() => { this.revealCard(card); }}
          />
        ));
      }
      rows.push(<div className="BoardRow" key={`row_${r}`}>{row}</div>);
    }
    const classes = ['Board'];
    switch (this.props.mode) {
      case BoardMode.Controller: classes.push('Board--controller'); break;
      case BoardMode.Viewer: classes.push('Board--viewer'); break;
      default: // do nothing!
    }
    return (
      <div className="BoardWrapper" onClick={() => this.highlightCard()}>
        <div className={classes.join(' ')}>{rows}</div>
      </div>
    );
  }

  getWord = (card: Coordinate): string => {
    return this.props.cards[this.coordinateToIndex(card)].word;
  }

  getRole = (card: Coordinate): Role => {
    return this.props.cards[this.coordinateToIndex(card)].role;
  }

  isRevealed = (card: Coordinate): boolean => {
    return this.props.cards[this.coordinateToIndex(card)].isRevealed;
  }

  isHighlighted = (card: Coordinate): boolean => {
    if (this.props.highlighted === undefined) { return false; }
    return this.props.highlighted.row === card.row && this.props.highlighted.col === card.col;
  }

  highlightCard = (card?: Coordinate) => {
    this.props.onHighlightCard(card);
  }

  revealCard = (card: Coordinate) => {
    this.props.onRevealCard(card);
  }

  private coordinateToIndex(coordinate: Coordinate): number {
    return coordinate.row * this.props.width + coordinate.col;
  }
}
