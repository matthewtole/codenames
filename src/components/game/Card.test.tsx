import * as React from 'react';
import { Card, CardProps } from './Card';
import * as renderer from 'react-test-renderer';
import { BoardMode } from './Board';
import { shallow } from 'enzyme';
import { Role } from '../../lib/types';

describe('Card', () => {
  const onReveal = jest.fn();
  const onHighlight = jest.fn();
  const stopPropagation = jest.fn();

  const defaultProps: CardProps = {
    word: 'test',
    role: Role.BLUE_SPY,
    boardMode: BoardMode.Viewer,
    revealed: false,
    highlighted: false,
    onReveal,
    onHighlight,
  };

  beforeEach(() => {
    onReveal.mockReset();
    onHighlight.mockReset();
  });

  it('renders correctly', () => {
    const tree = renderer.create(<Card {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call onHighlight when clicked and not highlighted', () => {
    const card = shallow(<Card {...defaultProps} />);
    card.simulate('click', { stopPropagation });
    expect(onReveal).not.toHaveBeenCalled();
    expect(onHighlight).toHaveBeenCalled();
  });

  it('should call onReveal when clicked and is highlighted', () => {
    const card = shallow(<Card {...defaultProps} highlighted={true} />);
    card.simulate('click', { stopPropagation });
    expect(onReveal).toHaveBeenCalled();
    expect(onHighlight).not.toHaveBeenCalled();
  });

  it('should call neither handler if revealed', () => {
    const card = shallow(<Card {...defaultProps} revealed={true} />);
    card.simulate('click', { stopPropagation });
    expect(onReveal).not.toHaveBeenCalled();
    expect(onHighlight).not.toHaveBeenCalled();
  });
});
