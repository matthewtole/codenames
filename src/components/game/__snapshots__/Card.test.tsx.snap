// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders correctly 1`] = `
<div
  className="Card Card--viewer"
  data-role="BLUE_SPY"
  onClick={[Function]}
>
  <div>
    test
  </div>
</div>
`;
